align_data/db/models.py

```
@@ -34,30 +41,63 @@ class Article(Base):
-
-    __id_fields = ['title', 'url']
+    __id_fields = ['url', 'title']
 
-    def __init__(self, *args, id_fields, **kwargs):
-        self.__id_fields = id_fields
+    def __init__(self, *args, id_fields=None, **kwargs):
         super().__init__(*args, **kwargs)
+        self.__id_fields = id_fields or self.__id_fields
+        if all(getattr(self, field, None) for field in self.__id_fields):
+            self.verify_fields()
+            id_string = self.generate_id_string()
+            self.id = hashlib.md5(id_string).hexdigest()
+        else:
+            raise ValueError(f"Entry is missing the following field(s): {','.join([missing for missing in self.__id_fields if not getattr(self, missing)])}")
 
     def __repr__(self) -> str:
-        return f"User(id={self.id!r}, name={self.title!r}, fullname={self.url!r})"
-
-    def generate_id_string(self):
+        return f"Article(id={self.id!r}, name={self.title!r}, fullname={self.url!r}, source={self.source!r}, source_type={self.source_type!r}, authors={self.authors!r}, date_published={self.date_published!r}, pinecone_update_required={self.pinecone_update_required!r}, pinecone_delete_required={self.pinecone_delete_required!r}, incomplete={self.incomplete!r})"
+    
+    def __eq__(self, other):
+        if not isinstance(other, Article):
+            return NotImplemented
+        return (
+            self.id == other.id and
+            self.title == other.title and
+            self.url == other.url and
+            self.source == other.source and
+            self.source_type == other.source_type and
+            self.authors == other.authors and
+            self.date_published == other.date_published and
+            self.text == other.text
+        )
+
+    def is_metadata_keys_equal(self, other):
+        if not isinstance(other, Article):
+            return NotImplemented
+        return not any(
+            key != 'entry_id' and getattr(self, key, None) != getattr(other, key, None) 
```

[NIT] this can fail if e.g 'bla' gets added to `PINECONE_METADATA_KEYS`, which should probably cause an error or something, rather than silently accept missing keys. This is a long way of saying that you don't need to check whether `key != 'entry_id'` for this to work, as `getattr(self, 'entry_id', None) == None`. So you can just do a basic `return all(getattr(self, key, None) == getattr(other, key, None)  for key in PINECONE_METADATA_KEYS)` if you don't care about nonexistent properties

---

Yeah I used to have it that way, and I guess it's more concise and the extra thing is unnecessary, but I thought it was more readable in case people wanted to understand/change/etc it in the future. It needs to be clear that entry_id is not relevant to this comparison method. Maybe a simple comment+code change to your version would be better?

---

totally up to you :D When I prefix something with `[NIT]` I'm mainly just adding small suggestions, which you're totally free to completely ignore